feat: yield protocol zapper studio integration

marcomariscal opened the pull request:
## Description

Integrate Yield Protocol into Zapper Studio:

Yield Protocol is a fixed-rate borrowing and lending protocol on Ethereum and Arbitrum.

There are three group id's that correspond to three different types of users of Yield: borrow (borrowers), lend (lenders), and pool (liquidity providers).

### Borrow

Borrowers open _vaults_ (analogous to maker), which are over-collateralized debt positions that have an associated maturity (i.e.: June 2022). Since these vaults are unique positions, a `token-fetcher` is not applicable and has not been created. For balances/user-specific data, the logic for fetching a user's associated vaults is included in the `balance-fetcher`.

### Lend
Lenders purchase fyTokens (ERC20) (i.e.: fyUSDC) that have a maturity (i.e.: June 2022), and are redeemable 1:1 for the underlying (i.e: USDC) at maturity. The value of an fyToken relative to its underlying can fluctuate before maturity, and so we estimate the "price" of an fyToken, to be able to use in the `balance-fetcher`.

### Pool
Users who want to provide liquidity use Yield strategies, which auto-roll to later maturities and mitigate manual user action. This is why we estimate the value of a strategy token (ERC20) relative to the pool's underlying (i.e.: USDC), to be used in the `balance-fetcher`.

## Checklist

- [x] I have followed the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] (optional) As a contributor, my Ethereum address/ENS is:
- [ ] (optional) As a contributor, my Twitter handle is:

## How to test?

1. The address `0x599aa2feaeec1c4caa33da6b7fbd0e6578953c96` should have a vault position on Ethereum, assuming the vault/borrow balance fetching was implemented correctly.
2. The address `0xd504170467f98f6bf01d90b2881d47adab302245` should have a lend position on Ethereum, assuming the lend token/balance fetching was implemented correctly.
3. The address `0x73a5dea042aeb13e9e9f81246281173f3ae3b0c3` should have a pool position on Ethereum and Arbitrum, assuming the pool/strategy fetching was implemented correctly.
4. The address `0x1793f656d03108ddbf56b5005a17142dc8fe1c40` should have a lend position on Arbitrum.
5. The address `0x35e057137060cd397bd82e3dff54beba480e7012` should have a vault on Arbitrum.

immasandwich commented:
Hey @marcomariscal ! Sorry I didn't have a chance to get to this today, I'll get you a review over the weekend for sure 🙏

immasandwich commented on src/apps/yield-protocol/arbitrum/yield-protocol.balance-fetcher.ts:
```diff
@@ -0,0 +1,209 @@
+
+        const positionBalance: ContractPositionBalance<YieldVaultContractPositionDataProps> = {
+          type: ContractType.POSITION,
+          address,
```
I think you want this address to be `0x16e25cf364cecc305590128335b8f327975d0560`. If I understand correctly, a user interacts with this address to open their vault, i.e.: [here](https://arbiscan.io/address/0x16e25cf364cecc305590128335b8f327975d0560?fromaddress=0x35e057137060cd397bd82e3dff54beba480e7012) is a list of transactions for the example address with vault balances you gave in your description.

marcomariscal replied:
cool! that's the `ladle` contract, which interacts with the `cauldron` contract, where the `cauldron` is the contract that holds the vaults. guessing i should use the cauldron then?

immasandwich replied:
As mentioned below, it'd be better to use the `ladle` address because it'd allow the Zapper backend to intelligently determine if a user has Yield Protocol vault open by checking if the user has a transaction to this address (the user interacts with the `ladle`, not the `cauldron`).

marcomariscal replied:
perfect! this has been updated.

immasandwich commented on src/apps/yield-protocol/arbitrum/yield-protocol.lend.token-fetcher.ts:
```diff
@@ -0,0 +1,180 @@
+
+        if (!underlyingToken) return null;
+
+        let pricePerShare = 0;
```
`pricePerShare` is the ratio between the price of the token and the price of the underlying token. In your case here, it should be `1` since they are redeemable 1:1.

We should document this better, and we will!

marcomariscal replied:
ah that makes total sense will update!

marcomariscal replied:
this has been updated

immasandwich commented on src/apps/yield-protocol/arbitrum/yield-protocol.pool.token-fetcher.ts:
```diff
@@ -0,0 +1,102 @@
+  '0xE779cd75E6c574d83D3FD6C92F3CBE31DD32B1E1',
+  '0xE7214Af14BD70F6AAC9c16B0c1Ec9ee1CcC7EFDA',
+  '0x92A5B31310a3ED4546e0541197a32101fCfBD5c8',
+  '0xDC705FB403DBB93Da1d28388bc1DC84274593c11',
```
Keep the addresses lowercased 🙏

marcomariscal replied:
this has been updated

immasandwich commented on src/apps/yield-protocol/arbitrum/yield-protocol.pool.token-fetcher.ts:
```diff
@@ -0,0 +1,102 @@
+
+        const displayProps: DisplayProps = {
+          label: `Yield ${underlyingToken.symbol} Strategy`,
+          secondaryLabel: `Automatic Roll on ${formatMaturity(maturity)}`,
```
You might want to have `secondaryLabel` as the price of the token, and `tertiaryLabel` for your maturity.

marcomariscal replied:
would like to keep as is because it is akin to a uni v2 pool (sort of), where the price isn't too meaningful for a user (they likely wouldn't act on that info).

immasandwich replied:
Got it 👍

immasandwich commented:
Really nice integration ser! I've left a few comments on the Arbitrum implementation that can also apply to the Ethereum implementation.

I would also recommend creating a `ContractPositionFetcher` for the borrow (vault) group with the address of the Ladle (i.e.: `0x16e25cf364cecc305590128335b8f327975d0560` on Arbitrum). You don't have to use it in your BalanceFetcher, but it'll enable Zapper API to do some more optimized balance fetching.

marcomariscal commented:
> Really nice integration ser! I've left a few comments on the Arbitrum implementation that can also apply to the Ethereum implementation.
> 
> I would also recommend creating a `ContractPositionFetcher` for the borrow (vault) group with the address of the Ladle (i.e.: `0x16e25cf364cecc305590128335b8f327975d0560` on Arbitrum). You don't have to use it in your BalanceFetcher, but it'll enable Zapper API to do some more optimized balance fetching.

Thanks so much @immasandwich! Will keep the balance fetcher as is then, and look to integrate a `ContractPositionFetcher` referencing the `Cauldron` (based on my note above emphasizing that the cauldron holds the vaults). Correct me if I'm wrong.

marcomariscal commented:
> > Really nice integration ser! I've left a few comments on the Arbitrum implementation that can also apply to the Ethereum implementation.
> > I would also recommend creating a `ContractPositionFetcher` for the borrow (vault) group with the address of the Ladle (i.e.: `0x16e25cf364cecc305590128335b8f327975d0560` on Arbitrum). You don't have to use it in your BalanceFetcher, but it'll enable Zapper API to do some more optimized balance fetching.
> 
> Thanks so much @immasandwich! Will keep the balance fetcher as is then, and look to integrate a `ContractPositionFetcher` referencing the `Cauldron` (based on my note above emphasizing that the cauldron holds the vault data). Correct me if I'm wrong.

Another quick question: so should the `ContractPositionFetcher` output show distinct debt/collateral asset pair data, or should it show something further granularity (or something else).

immasandwich commented:
Hello @marcomariscal ,

1. Regarding the `ContractPosition` address, use the address of the `ladle` (i.e.: the address that is used to enter the position). We'll document why in the future, but this enables Zapper to determine if a user has any vault positions open by checking if a user's address has every performed a transaction against the contract (and the user interacts with the `ladle`, not the `cauldron`).
2. As for the asset pair in the vault, it'd be good to have a position per asset pair (i.e.: supplied ETH and borrowed USDC, supplied ETH and borrowed DAI, whatever is supported by Yield protocol). This will be clearer in the future, but Studio developers will eventually be able enable Zapper users to enter these discrete positions through our UI.

immasandwich commented on src/apps/yield-protocol/arbitrum/yield-protocol.balance-fetcher.ts:
```diff
@@ -0,0 +1,210 @@
+        const displayProps: DisplayProps = {
+          label: `Yield Vault`,
+          secondaryLabel: `${getLabelFromToken(art)} Debt and ${getLabelFromToken(ilk)} Collateral`,
+          tertiaryLabel: formatMaturity(maturity),
```
I just remembered that for `ContractPosition` types, we currently only show the `label` and not the `secondaryLabel` or `tertiaryLabel` in the UI.

It might be crowded, but try to fit all the important information into the `label` for now, until we iterate to show the `secondaryLabel` and `tertiaryLabel` on the UI.

marcomariscal replied:
updated

immasandwich commented:
Looks great, let me know when you’re ready for the merge!

marcomariscal commented:
> > > Really nice integration ser! I've left a few comments on the Arbitrum implementation that can also apply to the Ethereum implementation.
> > > I would also recommend creating a `ContractPositionFetcher` for the borrow (vault) group with the address of the Ladle (i.e.: `0x16e25cf364cecc305590128335b8f327975d0560` on Arbitrum). You don't have to use it in your BalanceFetcher, but it'll enable Zapper API to do some more optimized balance fetching.
> > 
> > 
> > Thanks so much @immasandwich! Will keep the balance fetcher as is then, and look to integrate a `ContractPositionFetcher` referencing the `Cauldron` (based on my note above emphasizing that the cauldron holds the vault data). Correct me if I'm wrong.
> 
> Another quick question: so should the `ContractPositionFetcher` output show distinct debt/collateral asset pair data, or should it show something further granularity (or something else).



> Hello @marcomariscal ,
> 
> 1. Regarding the `ContractPosition` address, use the address of the `ladle` (i.e.: the address that is used to enter the position). We'll document why in the future, but this enables Zapper to determine if a user has any vault positions open by checking if a user's address has every performed a transaction against the contract (and the user interacts with the `ladle`, not the `cauldron`).
> 2. As for the asset pair in the vault, it'd be good to have a position per asset pair (i.e.: supplied ETH and borrowed USDC, supplied ETH and borrowed DAI, whatever is supported by Yield protocol). This will be clearer in the future, but Studio developers will eventually be able enable Zapper users to enter these discrete positions through our UI. You can see an example of this in `src/apps/maker/ethereum/maker.vault.contract-position-fetcher.ts`.

1. updated to use the ladle
2. updated to include `ContractPositionFetcher`(s), in which each position corresponds to a debt/collateral pair in Yield that

> Hello @marcomariscal ,
> 
> 1. Regarding the `ContractPosition` address, use the address of the `ladle` (i.e.: the address that is used to enter the position). We'll document why in the future, but this enables Zapper to determine if a user has any vault positions open by checking if a user's address has every performed a transaction against the contract (and the user interacts with the `ladle`, not the `cauldron`).
> 2. As for the asset pair in the vault, it'd be good to have a position per asset pair (i.e.: supplied ETH and borrowed USDC, supplied ETH and borrowed DAI, whatever is supported by Yield protocol). This will be clearer in the future, but Studio developers will eventually be able enable Zapper users to enter these discrete positions through our UI. You can see an example of this in `src/apps/maker/ethereum/maker.vault.contract-position-fetcher.ts`.

@immasandwich, updated with `ContractPositionFetchers` for borrow positions: we get distinct debt/collateral pair data if it's a valid pair. 

> Looks great, let me know when you’re ready for the merge!

@immasandwich, think we should be good!! 🤞

immasandwich commented:
Hey @marcomariscal , I'm seeing `Type `Query` has no field `vaultOwner`` for the Ethereum integration. Can you have a look at your API?

marcomariscal commented:
@immasandwich, currently in the process of updating our subgraphs (arbitrum is still syncing); will be pushing through a pr to address once arb is done syncing!
